contract_comm/currency/currency.go

```diff
@@ -216,128 +246,100 @@ func (cc *CurrencyComparator) Cmp(val1 *big.Int, currency1 *common.Address, val2
 		return val1.Cmp(val2)
 	}
 
-	// Below code block is basically evaluating this comparison:
-	// val1 * exchangeRate1.Denominator/exchangeRate1.Numerator < val2 * exchangeRate2.Denominator/exchangeRate2.Numerator
-	// It will transform that comparison to this, to remove having to deal with fractional values.
-	// val1 * exchangeRate1.Denominator * exchangeRate2.Numerator < val2 * exchangeRate2.Denominator * exchangeRate1.Numerator
-	leftSide := new(big.Int).Mul(val1, new(big.Int).Mul(exchangeRate1.Denominator, exchangeRate2.Numerator))
-	rightSide := new(big.Int).Mul(val2, new(big.Int).Mul(exchangeRate2.Denominator, exchangeRate1.Numerator))
-	return leftSide.Cmp(rightSide)
+	return exchangeRate1.CmpValues(val1, val2, exchangeRate2)
 }
 
-func Cmp(val1 *big.Int, currency1 *common.Address, val2 *big.Int, currency2 *common.Address) int {
-	// Short circuit if the fee currency is the same. nil currency => native currency
-	if (currency1 == nil && currency2 == nil) || (currency1 != nil && currency2 != nil && *currency1 == *currency2) {
-		return val1.Cmp(val2)
+func (cc *CurrencyManager) ToGold(amount *big.Int, currency *common.Address) (*big.Int, error) {
+	rate, err := cc.GetExchangeRate(currency)
+	if err != nil {
+		return nil, err
 	}
+	return rate.ToBase(amount), nil
+}
 
-	exchangeRate1, err1 := getExchangeRate(currency1)
-	exchangeRate2, err2 := getExchangeRate(currency2)
+func GetExchangeRate(currencyAddress *common.Address, header *types.Header, state vm.StateDB) (*ExchangeRate, error) {
+	if currencyAddress == nil {
+		return NoopExchangeRate(), nil
+	}
 
-	if err1 != nil || err2 != nil {
-		currency1Output := "nil"
-		if currency1 != nil {
-			currency1Output = currency1.Hex()
-		}
-		currency2Output := "nil"
-		if currency2 != nil {
-			currency2Output = currency2.Hex()
-		}
-		log.Warn("Error in retrieving exchange rate.  Will do comparison of two values without exchange rate conversion.", "currency1", currency1Output, "err1", err1, "currency2", currency2Output, "err2", err2)
-		return val1.Cmp(val2)
+	celoGoldAddress, err := contract_comm.GetRegisteredAddress(params.GoldTokenRegistryId, nil, nil)
+	if err == errors.ErrSmartContractNotDeployed || err == errors.ErrRegistryContractNotDeployed {
+		return nil, err
```

Comment:
This represents a change of behavior when comparing or converting currencies without using `ConvertToGold`, which was the only place where this was checked before. I'm concerned this could be a hard fork, and Its also very odd behavior because this will fail to get the CELO/CELO exchange rate when the registry is not deployed. We should likely put this check in the `ToGold` function and mark it as some unfortunate behavior to be removed at the next hardfork.

Reply:
good point, i didn't consider that. definitely unfortunate behaviour, i moved a couple of layers down to simplify/normalize the logic, but can be dangerous

I suspect it's not possible not use feeCurrency == celogGoldAddress, but i'm not 100% sure... i'll see how i can leave it as before

Reply:
I've checked on master, and `ConvertToGold` is only used on `tx_list` to check which of two txs pays more.

I don't think goldtoken address is actually a valid feeCurrency, so i doubt this is ever needed. And in any case, this is a shortcut. So i've removed the whole thing

Reply:
Removed this features, since `ConvertToGold` was only called on `txpool` and using a `feeCurrency=goldTokenAddress` is not (or shoul not)  be supported